File: java/src/test/java/algorithm/tests/geometry/PlanesTest.java
package algorithm.tests.geometry;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;

import algorithm.geometry.planes.Circle;
import algorithm.geometry.planes.Rectangle;
import algorithm.geometry.planes.Square;
import algorithm.tests.fixtures.Helper;
import algorithm.utils.errors.parameters.NegativeParameterError;
import algorithm.utils.errors.parameters.ZeroParameterError;

@Nested
@DisplayName("Plane Figures Test Suite")
public class PlanesTest {
    @Nested
    @DisplayName("Circle Test Suite")
    class CircleTest {
        @Test
        @DisplayName("Circle Negative or Zero Radius Test")
        void newCircle_negativeOrZeroRadius_throwsException() {
            Executable negativeCircle = () -> new Circle(-1);
            Executable zeroCircle = () -> new Circle(0);
            assertThrows(NegativeParameterError.class, negativeCircle);
            assertThrows(ZeroParameterError.class, zeroCircle);
        }

        @Test
        @DisplayName("Circle Radius Test")
        void circleRadius_returnsCorrectValue() throws Exception {
            Circle circle = new Circle(3.0);
            double actual = circle.getRadius();
            double expected = 3.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Circle Area Test")
        void circleArea_returnsCorrectValue() throws Exception {
            Circle circle = new Circle(3.0);
            double actual = circle.area();
            double expected = Math.PI * Math.pow(3.0, 2);
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Circle Perimeter Test")
        void circlePerimeter_returnsCorrectValue() throws Exception {
            Circle circle = new Circle(3.0);
            double actual = circle.perimeter();
            double expected = 2 * Math.PI * 3.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }
    }

    @Nested
    @DisplayName("Rectangle Test Suite")
    class RectangleTest {
        @Test
        @DisplayName("Rectangle Negative or Zero Length Test")
        void newRectangle_negativeOrZeroLength_throwsException() {
            Executable negativeRectangle = () -> new Rectangle(-1, 2);
            Executable zeroRectangle = () -> new Rectangle(0, 2);
            assertThrows(NegativeParameterError.class, negativeRectangle);
            assertThrows(ZeroParameterError.class, zeroRectangle);
        }

        @Test
        @DisplayName("Rectangle Negative or Zero Width Test")
        void newRectangle_negativeOrZeroWidth_throwsException() {
            Executable negativeRectangle = () -> new Rectangle(1, -2);
            Executable zeroRectangle = () -> new Rectangle(1, 0);
            assertThrows(NegativeParameterError.class, negativeRectangle);
            assertThrows(ZeroParameterError.class, zeroRectangle);
        }

        @Test
        @DisplayName("Rectangle Length Test")
        void rectangleLength_returnsCorrectValue() throws Exception {
            Rectangle rectangle = new Rectangle(3.0, 4.0);
            double actual = rectangle.getWidth();
            double expected = 3.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Rectangle Width Test")
        void rectangleWidth_returnsCorrectValue() throws Exception {
            Rectangle rectangle = new Rectangle(3.0, 4.0);
            double actual = rectangle.getHeight();
            double expected = 4.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Rectangle Area Test")
        void rectangleArea_returnsCorrectValue() throws Exception {
            Rectangle rectangle = new Rectangle(3.0, 4.0);
            double actual = rectangle.area();
            double expected = 3.0 * 4.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Rectangle Perimeter Test")
        void rectanglePerimeter_returnsCorrectValue() throws Exception {
            Rectangle rectangle = new Rectangle(3.0, 4.0);
            double actual = rectangle.perimeter();
            double expected = 2 * (3.0 + 4.0);
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }
    }

    @Nested
    @DisplayName("Square Test Suite")
    class SquareTest {
        @Test
        @DisplayName("Square Negative or Zero Side Test")
        void newSquare_negativeOrZeroSide_throwsException() {
            Executable negativeSquare = () -> new Square(-1);
            Executable zeroSquare = () -> new Square(0);
            assertThrows(NegativeParameterError.class, negativeSquare);
            assertThrows(ZeroParameterError.class, zeroSquare);
        }

        @Test
        @DisplayName("Square Side Test")
        void squareSide_returnsCorrectValue() throws Exception {
            Square square = new Square(3.0);
            double actual = square.getSide();
            double expected = 3.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Square Area Test")
        void squareArea_returnsCorrectValue() throws Exception {
            Square square = new Square(3.0);
            double actual = square.area();
            double expected = Math.pow(3.0, 2);
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Square Perimeter Test")
        void squarePerimeter_returnsCorrectValue() throws Exception {
            Square square = new Square(3.0);
            double actual = square.perimeter();
            double expected = 4 * 3.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }
    }
}

File: rust/tests/geometry/planes_test.rs
use algorithm::geometry::Plane;
use algorithm::geometry::planes::{Circle, Rectangle, Square};
use algorithm::utils::errors::parameters::{NegativeParameterError, ZeroParameterError};

use crate::fixtures::helper;

#[cfg(test)]
mod circle_test {
    use super::*;

    #[test]
    fn circle_new_negative_or_zero_radius_returns_error() {
        let negative_circle = Circle::new(-1.0);
        let zero_circle = Circle::new(0.0);
        assert!(matches!(negative_circle, Err(e) if e.is::<NegativeParameterError>()));
        assert!(matches!(zero_circle, Err(e) if e.is::<ZeroParameterError>()));
    }

    #[test]
    fn circle_radius_returns_correct_value() {
        let circle = Circle::new(3.0).unwrap();
        let actual = circle.radius();
        let expected = 3.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn circle_area_returns_correct_value() {
        let circle = Circle::new(3.0).unwrap();
        let actual = circle.area();
        let expected = 28.274333882308138;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn circle_perimeter_returns_correct_value() {
        let circle = Circle::new(3.0).unwrap();
        let actual = circle.perimeter();
        let expected = 18.84955592153876;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }
}

#[cfg(test)]
mod rectangle_test {
    use super::*;
    #[test]
    fn rectangle_new_negative_or_zero_width_returns_error() {
        let negative_rectangle = Rectangle::new(-1.0, 2.0);
        let zero_rectangle = Rectangle::new(0.0, 2.0);
        assert!(matches!(negative_rectangle, Err(e) if e.is::<NegativeParameterError>()));
        assert!(matches!(zero_rectangle, Err(e) if e.is::<ZeroParameterError>()));
    }

    #[test]
    fn rectangle_new_negative_or_zero_height_returns_error() {
        let negative_rectangle = Rectangle::new(2.0, -1.0);
        let zero_rectangle = Rectangle::new(2.0, 0.0);
        assert!(matches!(negative_rectangle, Err(e) if e.is::<NegativeParameterError>()));
        assert!(matches!(zero_rectangle, Err(e) if e.is::<ZeroParameterError>()));
    }

    #[test]
    fn rectangle_width_returns_correct_value() {
        let rectangle = Rectangle::new(4.0, 5.0).unwrap();
        let actual = rectangle.width();
        let expected = 4.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn rectangle_height_returns_correct_value() {
        let rectangle = Rectangle::new(4.0, 5.0).unwrap();
        let actual = rectangle.height();
        let expected = 5.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn rectangle_area_returns_correct_value() {
        let rectangle = Rectangle::new(4.0, 5.0).unwrap();
        let actual = rectangle.area();
        let expected = 20.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn rectangle_perimeter_returns_correct_value() {
        let rectangle = Rectangle::new(4.0, 5.0).unwrap();
        let actual = rectangle.perimeter();
        let expected = 18.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }
}

#[cfg(test)]
mod square_test {
    use super::*;

    #[test]
    fn square_new_negative_or_zero_side_returns_error() {
        let negative_square = Square::new(-1.0);
        let zero_square = Square::new(0.0);
        assert!(matches!(negative_square, Err(e) if e.is::<NegativeParameterError>()));
        assert!(matches!(zero_square, Err(e) if e.is::<ZeroParameterError>()));
    }

    #[test]
    fn square_side_returns_correct_value() {
        let square = Square::new(4.0).unwrap();
        let actual = square.side();
        let expected = 4.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn square_area_returns_correct_value() {
        let square = Square::new(4.0).unwrap();
        let actual = square.area();
        let expected = 16.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn square_perimeter_returns_correct_value() {
        let square = Square::new(4.0).unwrap();
        let actual = square.perimeter();
        let expected = 16.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }
}
